subprojects/core/src/main/java/org/gradle/api/internal/provider/DefaultListProperty.java

```diff
@@ -17,21 +17,56 @@
 package org.gradle.api.internal.provider;
 
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
 import org.gradle.api.Transformer;
 import org.gradle.api.provider.ListProperty;
 import org.gradle.api.provider.Provider;
 import org.gradle.internal.Cast;
 
 import javax.annotation.Nullable;
+import java.util.Collection;
 import java.util.List;
+import java.util.concurrent.Callable;
 
 class DefaultListProperty<T> implements PropertyInternal<List<T>>, ListProperty<T> {
     private static final Provider<ImmutableList<Object>> EMPTY_LIST = Providers.of(ImmutableList.of());
-    private Provider<? extends List<T>> provider = Cast.uncheckedCast(EMPTY_LIST);
+    private Provider<? extends Collection<T>> provider = Cast.uncheckedCast(EMPTY_LIST);
 
     public DefaultListProperty(Class<T> elementType) {
     }
 
+    @Override
+    public void add(final T element) {
+        addAll(Providers.of(ImmutableList.of(element)));
+    }
+
+    @Override
+    public void add(final Provider<? extends T> providerOfElement) {
+        addAll(new DefaultProvider<List<T>>(new Callable<List<T>>() {
+            @Override
+            public List<T> call() throws Exception {
+                return ImmutableList.of(providerOfElement.get());
+            }
+        }));
+    }
+
+    @Override
+    public void addAll(final Provider<? extends Collection<T>> providerOfElements) {
+        if (provider == Cast.uncheckedCast(EMPTY_LIST)) {
```

Comment:
Is this unchecked cast necessary?  When I remove it, I don't think I see warnings in IDEA.

Reply:
We may also want to put this in a private method with a good name to explain what we're checking.

Reply:
Good point, I will change that.